snapcraft/storeapi/assertions.py

```diff
@@ -0,0 +1,262 @@
+            cmdline += ['-k', self._signing_key]
+        snap_sign = subprocess.Popen(
+            cmdline, stdin=subprocess.PIPE, stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
+        data = json.dumps(self.assertion).encode('utf8')
+        assertion, err = snap_sign.communicate(input=data)
+        if snap_sign.returncode != 0:
+            err = err.decode('ascii', errors='replace')
+            raise errors.StoreAssertionError(
+                snap_name=self._snap_name,
+                endpoint=self._assertion_type, error=err)
+
+        self.signed_assertion = assertion
+        return assertion
+
+    def push(self, force=False):
+        """Push the assertion to the store, signing if necessary.
+
+        :param bool force: if True, ignore any conflict with revoked developers
+                           and the snap revisions it would invalidate.
+        :returns: None
+        """
+        if not self.signed_assertion:
+            self.sign()
+        self._store_client.push_assertion(
+            self.snap_id, self.signed_assertion, self._assertion_type,
+            force=force)
+
+
+class DeveloperAssertion(_BaseAssertion):
+    """Implementation of a developer assertion.
+    The assertion is used to enable collaboration for a given snap_name by
+    means of operating agains the snap-id's developer assertion store endpoint.
+
+    The assertion content has the following structure
+    {
+        'type': 'snap-developer',
+        'authority-id': '<account_id of the publisher>',
+        'publisher-id': '<account_id of the publisher>',
+        'snap-id': '<snap-id>',
+        'developers': [{
+            'developer-id': '<account-id>',
+            'since': '2017-02-10T08:35:00.390258Z'
+           },{
+            'developer-id': '<account-id>',
+            'since': '2017-02-10T08:35:00.390258Z',
+            'until': '2018-02-10T08:35:00.390258Z'
+           }],
+        }
+
+    """
+
+    _assertion_type = 'developers'
+
+    def new_assertion(self, *, developers):
+        """Create a new assertion with developers based out of the current one.
+
+        The new assertion has its assertion's authority-id normalized to the
+        assertion's publisher_id and the assertion's revision increaded by 1.
+
+        :param list developers: a list of a dictionary developers holding the
+                                keys: developer_id (mandatory), since
+                                (mandatory) and until (optional).
+        :returns: a new assertion based out of the current assertion with the
+                  provided developers list.
+        :rtype: DeveloperAssertion
+        """
+        new_assertion = deepcopy(self.assertion)
+        new_assertion['developers'] = developers
+
+        # The revision should be incremented, to avoid `invalid-revision`
+        # errors.
+        new_assertion['revision'] = str(int(
+            self.assertion.get('revision', '0'))+1)
```

Comment:
By convention, the first revision of an assertion is zero so you could probably pass `get` a '-1' default ... but things will still work like this.